Network Namespaces

nixbitcoin opened the pull request:
Network Namespaces are a significant security improvement, mitigating the bind race leak security vulnerability and unnecessary communications between nix-bitcoin services. Network namespaces round-off our service isolation to cover all levels of inter-service interaction.

Special thank you to @erikarvstedt for generalizing the `makeNetnsServices` :rocket:

Kixunil commented:
Regarding 4) I solved a similar problem for `lncli` by writing [a wrapper script](https://github.com/Kixunil/cryptoanarchy-deb-repo-builder/blob/master/lnd/xlncli) and named it `lncli` using `update-alternatives`. I plan to do the same for `bitcoin-cli` Maybe a similar approach can work here?

mmilata commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,257 @@
+  isEnabled = x: config.services.${x}.enable;
+
+  ip = "${pkgs.iproute}/bin/ip";
+  iptables = "${pkgs.iptables}/bin/iptables";
```
Consider using `config.networking.firewall.package` instead for nftables. (Though if nixos/nixpkgs#81172 should make this uneccessary.)

nixbitcoin replied:
Is nftables better than iptables for what we're trying to do?

mmilata replied:
Sorry, I meant to write "... instead for *compatibility with* nftables". Using `config.networking.firewall.package` instead of a concrete package would make it work regardless of whether the system is set to use iptables or nftables.

IMO nftables is the future but both work fine for this use case.

nixbitcoin replied:
See `fixup! architecture: pkgs.iptables -> config.networking.firewall.package`

mmilata commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,257 @@
+    networking.firewall.interfaces.br0.allowedTCPPorts = [ 9050 ];
+    boot.kernel.sysctl."net.ipv4.ip_forward" = true;
+
+    nix-bitcoin.netns-isolation.services = {
```
For the sake of modularity it might be preferable to define the implementation in this file and then set the concrete mapping and other parameters elsewhere, e.g. in secure-node.nix?

nixbitcoin replied:
@jonasnick and I talked about restructuring this file like so: parameters in each individual module, mapping in `secure-node.nix` and general implementation in `netns-chef.nix`. What are your thoughts on this approach?

mmilata replied:
Sounds reasonable!

nixbitcoin commented:
> Regarding 4) I solved a similar problem for `lncli` by writing [a wrapper script](https://github.com/Kixunil/cryptoanarchy-deb-repo-builder/blob/master/lnd/xlncli) and named it `lncli` using `update-alternatives`. I plan to do the same for `bitcoin-cli` Maybe a similar approach can work here?

We've used wrappers before, check out [this](https://github.com/fort-nix/nix-bitcoin/blob/master/modules/bitcoind.nix#L238). The problem is that `ip netns exec` requires elevated privileges, which I don't think is a good idea to give the `operator` user. If the `operator` user can run `ip netns exec` with root privileges, he can run any following command with root privileges.

I've been thinking about how to solve 4. and I think the best idea is to create a separate `nb-operator` namespace with access to all other namespaces. All the user scripts will automatically execute with the right flags, like `-rpcconnect` for `bitcoin-cli`. When the operator user ssh's into the box he's automatically placed into a shell inside the `nb-operator` network namespace. Thoughts?

nixbitcoin commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,257 @@
+
+    nix-bitcoin.netns-isolation.services = {
+      bitcoind = {
+        id = 12;
```
Why would you want to do that? It makes everything simpler IMO to just have hardcoded addresses in a changeable addressblock.

mmilata commented:
What about [setcap wrapper](https://nixos.org/nixos/options.html#security.wrappers)? Namespaces require CAP_SYS_ADMIN which is [pretty much root](https://lwn.net/Articles/486306/) but I guess it's marginally better than the sudo rule and the capability won't be inherited by the program passed to `ip netns exec` as an argument?

erikarvstedt commented:
Please rename `netns-chef.nix` -> `netns-isolation.nix`. chef is amusing in the context of onions, but it's inappropriate here.

jonasnick commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,146 @@
+    enable = mkEnableOption "netns isolation";
+
+    addressblock = mkOption {
+      type = types.str;
```
Why not `type.u8`?

nixbitcoin replied:
Because we literally only change the N in `169.254.N.0/24`

jonasnick commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,146 @@
+      default = {};
+      type = types.attrsOf (types.submodule {
+        options = {
+          id = mkOption { type = types.int; };
```
Description would be helpful. "must be unique", "must not be 10"

nixbitcoin replied:
See `fixup! id option: add description`

jonasnick commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,146 @@
+
+  config = mkIf cfg.enable {
+    # Prerequisites
+    networking.dhcpcd.extraConfig = "noipv4ll";
```
Would be nice to have a comment here explaining why this is needed. Couldn't we use dhcpcd's `denyinterfaces` option to ignore the interfaces altogether?

nixbitcoin replied:
See `fixup! dhcpcd: noipv4ll -> denyinterfaces`

jonasnick commented on modules/netns-chef.nix:
```diff
@@ -0,0 +1,146 @@
+  config = mkIf cfg.enable {
+    # Prerequisites
+    networking.dhcpcd.extraConfig = "noipv4ll";
+    services.tor.client.socksListenAddress = "169.254.${toString cfg.addressblock}.10:9050";
```
Better `let bridgeIp = "...";` because it's used quite a few times in this file.

nixbitcoin replied:
See `fixup! let: add bridgeIp` and `fixup! let: add operatorIp`

jonasnick commented on modules/bitcoind.nix:
```diff
@@ -117,6 +119,27 @@ in {
         };
       };
+      rpcbind = mkOption {
+        type = types.lines;
```
Would be more consistent and simpler if this was just a list of ip addresses (same with `rpcallowip` and in the liquid module)

nixbitcoin replied:
See `fixup! liquid: make rpcbind and rpcallowip lists of ip addresses` and `fixup! bitcoind: make rpcbind and rpcallowip lists of ip addresses`

jonasnick commented:
It would be helpful to document how this works and in particular how this can be customized to allow access through clearnet or wireguard interfaces. As an example here's the diagram I sketched while trying to understand what is happening.

![crappy-diagram](https://user-images.githubusercontent.com/2582071/83659599-42ffa280-a5b3-11ea-9879-699991a47cb2.jpg)

nixbitcoin commented:
I force pushed a new and improved HEAD that includes lnd, an operator namespace, and several fixes. It also removes the clightning autolisten option, as documented in the commit.

These todos remain:

I've tried implementing these and have gotten errors that I can't solve (please help @erikarvstedt @jonasnick):
1. Restructure the custom netns configs so they are only set when the individual services are actually enabled.
2. Make `bitcoind-import-banlist` run in the `nb-bitcoind namespace`.

I need your conceptual input on these:
1. Think about placing custom netns configs in each individual module, mapping in `secure-node.nix` and general implementation in `netns-isolation.nix`.
2. Drop the operator user into a shell inside `nb-operator` and give him aliases that include flags like `rpcconnect`. Something like `ip netns exec nb-operator sudo -u operator bash` should do.
3. Mitigate or document LND's breaking changes.
4. Find a better way to amend the LND cert than hardcoding the ip address.
5. Update the testing framework to test with and without network namespaces.

I can take care of these but they depend on the above:
1. Document clearnet and wireguard access.

nixbitcoin commented:
I got rid of the operator shell and replaced it with custom cli scripts that use Jonas Nick's `netns-exec` to execute `bitcoin-cli`, `lncli`, and `elements-cli` in the appropriate network namespaces. Operator can use `netns-exec` because of a `security.wrappers` program with `cap_sys_admin`.

I also rewrote the `test_script.py` to use netns-isolation. If we want to test without netns-isolation we need to write a second `test_script.py`. Tests run nicely now.

Only remaining issues are the breaking changes to `lnd` (hardcoded ip address in cert and new hidden service).

Very close to mergeable IMO.

nixbitcoin commented:
Fixed some nits.

nixbitcoin commented on test/test-script.py:
```diff
@@ -38,13 +38,13 @@ def assert_running(unit):
 
 assert_running("electrs")
-machine.wait_for_open_port(4224)  # prometeus metrics provider
+machine.wait_until_succeeds("ip netns exec nb-electrs nc -z localhost 4224")
```
Comment missing.

nixbitcoin replied:
Fixed in a2dc3da63530a8793e1e689927b226e15e062cd1

nixbitcoin commented on test/test-script.py:
```diff
@@ -58,31 +58,49 @@ def assert_running(unit):
 
+### Security tests
+
+machine.fail("ip netns exec nb-bitcoind ping -c 1 169.254.1.17")
```
Fix tests by reformatting, setting ping deadline, and also testing successful pings.

nixbitcoin replied:
Fixed in a2dc3da63530a8793e1e689927b226e15e062cd1

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+  ip = "${pkgs.iproute}/bin/ip";
+  iptables = "${config.networking.firewall.package}/bin/iptables";
+
+  bridgeIp = "169.254.${toString cfg.addressblock}.10";
```
Remove `toString`

nixbitcoin replied:
Actually do need `toString`

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+            ${ip} link add name br0 type bridge
+            ${ip} link set br0 up
+            ${ip} addr add ${bridgeIp}/24 brd + dev br0
+            ${iptables} -w -t nat -A POSTROUTING -s 169.254.${toString cfg.addressblock}.0/24 -j MASQUERADE
```
Remove toString

nixbitcoin replied:
Actually do need `toString`

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+            ${iptables} -w -t nat -A POSTROUTING -s 169.254.${toString cfg.addressblock}.0/24 -j MASQUERADE
+          '';
+          preStop = ''
+            ${iptables} -w -t nat -D POSTROUTING -s 169.254.${toString cfg.addressblock}.0/24 -j MASQUERADE
```
Remove toString

nixbitcoin replied:
Actually do need `toString`

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+      services.clightning = mkIf config.services.clightning.enable {
+        host = "${netns.clightning.address}";
+        bitcoin-rpcconnect = "${netns.bitcoind.address}";
+        bind-addr = "${netns.clightning.address}:${toString config.services.clightning.onionport}";
```
Remove toString

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+      services.electrs = mkIf config.services.electrs.enable {
+        host =  if config.services.electrs.TLSProxy.enable then "${toString netns.nginx.address}" else "${toString netns.electrs.address}";
+        address = "${toString netns.electrs.address}";
+        daemonrpc = "${toString netns.bitcoind.address}:${toString config.services.bitcoind.rpc.port}";
```
Remove toString's

nixbitcoin replied:
Fixed in b51b7d6530e6f123e63da78442e70c03e7babeb3

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+      };
+
+      # lightning-charge: Custom netns configs
+      services.lightning-charge.host = mkIf config.services.lightning-charge.enable "${toString netns.lightning-charge.address}";
```
Remove toString

nixbitcoin replied:
Fixed in aea2de95045a7a5071815e43c24b7a261c344b93

nixbitcoin commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+
+      # spark-wallet: Custom netns configs
+      services.spark-wallet = mkIf config.services.spark-wallet.enable {
+        host = "${netns.spark-wallet.address}";
```
Remove dollar_curly

nixbitcoin replied:
Fixed in ac22262bf762fdc2b5b49f7da7b8a24e55024287

jonasnick commented on modules/bitcoind.nix:
```diff
@@ -68,6 +70,13 @@ in {
         default = "/var/lib/bitcoind";
         description = "The data directory for bitcoind.";
       };
+      host = mkOption {
```
I feel like we don't really need these host options. In some modules (like bitcoind) they aren't even used. As far as I can see they are mainly used for the hidden service `toHost` option. For clightning for example we can directly use bind-addr. For lnd we can keep host (or rename to listen, because that's how the option is called in lnd). For bitcoind we can add a `bind` option (we should do that anyway) which if set will result in an additional line `bind=${...}` in the bitcoind config.

nixbitcoin replied:
Fixed where possible. Nanopos, lightning-charge and spark-wallet actually have a host option, so naming is correct. It's needed in electrs for now to direct traffic to the nginx netns, can be removed when #203 is merged.

jonasnick commented on modules/netns-isolation.nix:
```diff
@@ -0,0 +1,296 @@
+              ${ip} link set br-${vethName} master br0
+              ${ipNetns} route add default via ${bridgeIp}
+              ${netnsIptables} -w -P INPUT DROP
+              ${netnsIptables} -w -A INPUT -s 127.0.0.1,${bridgeIp} -j ACCEPT
```
It should be possible to allow the `ipNetns` to connect to `ipNetns`. In order to allow that I think we need to add `${ipNetns}` here and in the `OUTPUT` rule below. That allows simplifying the tests by testing connectivity from the same container. Also we can remove additional connection from clightning to spark in the tests.

nixbitcoin replied:
Fixed in 43a34be11cf79d252cbc0b846b12623b61c4fdbb

nixbitcoin commented on modules/bitcoind.nix:
```diff
@@ -233,12 +256,22 @@ in {
       cli = mkOption {
         type = types.package;
-        readOnly = true;
         default = pkgs.writeScriptBin "bitcoin-cli" ''
```
Fixed in dda5047e92b41bb0bc076971e9d6005ec3af1ced

nixbitcoin commented:
Rebased and fixed @jonasnick comments. 

<details>
  <summary>a2dc3da63530a8793e1e689927b226e15e062cd1 notes</summary>

- Disabled module by default
- Inserted ${ipNetns} firewall rule
- Removed unnecessary toStrings
- Removed unnecessary DOLLAR_CURLY's
- Defaulted bitcoin cli to cli-nonetns-exec
- Got rid of host options and used service-specific options
- Added `rpclisten` and `restlisten` options of type `types.listOf types.str` to lnd module
- Fixed missing comment in test_script.py
- Simplified spark-wallet test by testing from same container
- Set ping deadline
- Implemented must succeed pings
- Implemented netns-exec tests
  - That it drops capabilities
  - That it can't execute in bad namespaces
</details>

Running tests with and without `netns-isolation` requires creating something like
```
nodes =
    { machine1 =
        { config, pkgs, ... }: { … };
      machine2 =
        { config, pkgs, ... }: { … };
      …
    };
```
where `nix-bitcoin.netns-isolation.enable = mkForce true;` for one of the machines. `test-script.py` would have to be rewritten to perform tests on both machines concurrently. I think @erikarvstedt is the best person to implement this elegantly.

jonasnick commented on test/test-script.py:
```diff
@@ -58,31 +66,67 @@ def assert_running(unit):
+
+# Positive ping tests
+machine.succeed(
+    "ip netns exec nb-bitcoind ping -c 1 -w 1 169.254.1.12 && \
```
Can you define constants for these addresses? That would give them a name and make it easy to see what they're referring to.

nixbitcoin replied:
Fixed in 9d7181b5aaadf326728a3a5c9b255f0cae36598c

jonasnick commented on test/test-script.py:
```diff
@@ -58,31 +66,67 @@ def assert_running(unit):
+)
+
+# netns-exec tests
+machine.fail("netns-exec nb-electrs ip a")
```
Can you add a comment to explain what this does and why this must fail?

jonasnick replied:
And how about checking that netns-exec can not be executed by users that are not operator?

nixbitcoin replied:
Both fixed in 9d7181b5aaadf326728a3a5c9b255f0cae36598c

jonasnick commented on test/test-script.py:
```diff
@@ -58,31 +66,67 @@ def assert_running(unit):
+machine.fail("netns-exec nb-electrs ip a")
+assert_matches_exactly(
+    "su operator -c 'netns-exec nb-bitcoind capsh --print | grep Current | tr -d \"\\n\"'",
+    "Current: =",
```
Can't you match against "Current: =\n" instead of using `tr`?

nixbitcoin replied:
Fixed in a3aaff192b1988064f22ad77a25514c2e535ee1f

jonasnick commented on .travis.yml:
```diff
@@ -14,7 +14,8 @@ env:
   jobs:
-  - TestModules=1 STABLE=1
+  - DefaultTestModules=1 STABLE=1
+  - NetnsTestModules=1 STABLE=1
```
Perhaps simpler to directly define a `SCENARIO=` variable that is either "default" or "withnetns".

nixbitcoin replied:
Fixed in 8a7c9b2839b1628679c14091a08bf36d549ae0d1
Much better

jonasnick commented on test/scenarios/default.py:
```diff
@@ -1,32 +1,3 @@
-
-# Don't execute the following test suite when this script is running in interactive mode
-if "is_interactive" in vars():
-    raise Exception()
```
Seems like this is missing from `lib.py` now.

nixbitcoin replied:
Fixed in d220ab8e55c5f0536d2bf57aca7bdf6a3bb995a1

nixbitcoin commented:
Tests now run with multiple different configurations called `scenarios`. Travis automatically tests all specified scenarios. `./run.tests.sh --scenario <scenario> debug` confirmed working.

jonasnick commented on test/scenarios/withnetns.py:
```diff
@@ -0,0 +1,139 @@
+
+# Positive ping tests (non-exhaustive)
+machine.succeed(
+    "bitcoindip='169.254.1.12' && \
```
Better to define these vars in python. Then you only have to do it once. See  https://github.com/jonasnick/nix-bitcoin/commits/withnetns-test-fix

nixbitcoin replied:
Also replaced the ping commands and netns ip's in the entire script with python vars. ceaf1890fdb7a5b8c6d63436c37d193bc80db9f9

nixbitcoin commented:
Had to make bind dependent on listen with `${optionalString cfg.listen "bind=${cfg.bind}"}` in `liquidd` and `bitcoind`. Tests were failing with `Error: Cannot set -bind or -whitebind together with -listen=0`.

nixbitcoin commented:
Moved `nix-bitcoin.netns-isolation.enable` to `configuration.nix`

nixbitcoin commented:
Fixed final nits
